src/client/datascience/jupyter/kernels/cellExecution.ts

```diff
@@ -252,112 +225,51 @@ export class CellExecution {
      */
     private dispose() {
         traceCellMessage(this.cell, 'Execution disposed');
-        this.activeExecution = undefined;
         this.disposables.forEach((d) => d.dispose());
         const deferred = CellExecution.cellsCompletedForTesting.get(this.cell);
         if (deferred) {
             deferred.resolve();
         }
     }
 
-    private async interruptExecution(
-        session: IJupyterSession,
-        execution: Promise<void>,
-        timeoutMS: number
-    ): Promise<InterruptResult> {
-        // Create a deferred promise that resolves if we have a failure
-        const restarted = createDeferred<boolean>();
-
-        // Listen to status change events so we can tell if we're restarting
-        const restartHandler = (e: ServerStatus) => {
-            if (e === ServerStatus.Restarting) {
-                // We restarted the kernel.
-                traceWarning('Kernel restarting during interrupt');
-
-                // Indicate we restarted the race below
-                restarted.resolve(true);
-            }
-        };
-        const restartHandlerToken = session.onSessionStatusChanged(restartHandler);
-
-        // Start our interrupt. If it fails, indicate a restart
-        session.interrupt(timeoutMS).catch((exc) => {
-            traceWarning(`Error during interrupt: ${exc}`);
-            restarted.resolve(true);
-        });
-
-        try {
-            // Wait for all of the pending cells to finish or the timeout to fire
-            const result = await waitForPromise(Promise.race([execution, restarted.promise]), timeoutMS);
-
-            // See if we restarted or not
-            if (restarted.completed) {
-                return InterruptResult.Restarted;
-            }
-
-            if (result === null) {
-                // We timed out. You might think we should stop our pending list, but that's not
-                // up to us. The cells are still executing. The user has to request a restart or try again
-                return InterruptResult.TimedOut;
-            }
-
-            // Indicate the interrupt worked.
-            return InterruptResult.Success;
-        } catch (exc) {
-            // Something failed. See if we restarted or not.
-            if (restarted.completed) {
-                return InterruptResult.Restarted;
-            }
-
-            // Otherwise a real error occurred.
-            throw exc;
-        } finally {
-            restartHandlerToken.dispose();
-        }
-    }
-    private handleKernelRestart(kernel: IKernel) {
```

Comment:
Moved to `KernelExecution`, this was `CellExecution` never has to cancel itself.
Moved a lot of these out to kernel execution, cell execution should never cancel itself, it should be told to cancel.
Much clear, removes need to pass in `IKernel` as well.

Reply:
I like these moves. Makes it more clear. I remember being confused some on previous reviews by the cancellation / interrupt flow.